rcl/src/rcl/type_version_hash.c

```diff
@@ -0,0 +1,257 @@
+    individual_type_description->type_name.size);
+
+  // fields
+  EMIT_KEY("fields");
+  {
+    FAILFAST(yaml_sequence_start_event_initialize(&event, NULL, NULL, 1, YAML_FLOW_SEQUENCE_STYLE));
+    EMIT();
+    for (size_t i = 0; i < individual_type_description->fields.size; i++) {
+      FAILFAST(emit_field(emitter, &individual_type_description->fields.data[i]));
+    }
+    FAILFAST(yaml_sequence_end_event_initialize(&event));
+    EMIT();
+  }
+
+  FAILFAST(yaml_mapping_end_event_initialize(&event));
+  EMIT();
+  return 1;
+}
+
+static int emit_type_description(
+  yaml_emitter_t * emitter,
+  const type_description_interfaces__msg__TypeDescription * type_description)
+{
+  yaml_event_t event;
+
+  FAILFAST(yaml_mapping_start_event_initialize(&event, NULL, NULL, 1, YAML_FLOW_MAPPING_STYLE));
+  EMIT();
+
+  EMIT_KEY("type_description");
+  FAILFAST(emit_individual_type_description(emitter, &type_description->type_description));
+
+  EMIT_KEY("referenced_type_descriptions");
+  {
+    FAILFAST(yaml_sequence_start_event_initialize(&event, NULL, NULL, 1, YAML_FLOW_SEQUENCE_STYLE));
+    EMIT();
+    for (size_t i = 0; i < type_description->referenced_type_descriptions.size; i++) {
+      FAILFAST(
+        emit_individual_type_description(
+          emitter, &type_description->referenced_type_descriptions.data[i]));
+    }
+    FAILFAST(yaml_sequence_end_event_initialize(&event));
+    EMIT();
+  }
+
+  FAILFAST(yaml_mapping_end_event_initialize(&event));
+  EMIT();
+  return 1;
+}
+
+rcl_ret_t
+rcl_type_description_to_hashable_json(
+  const type_description_interfaces__msg__TypeDescription * type_description,
+  rcutils_char_array_t * output_repr)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(type_description, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(output_repr, RCL_RET_INVALID_ARGUMENT);
+
+  yaml_emitter_t emitter;
+  yaml_event_t event;
+
+  if (!yaml_emitter_initialize(&emitter)) {goto error;}
+  // Disable line breaks based on line length
+  yaml_emitter_set_width(&emitter, -1);
+  // Circumvent EOF line break by providing invalid break style
+  yaml_emitter_set_break(&emitter, -1);
+
+  yaml_emitter_set_output(&emitter, yaml_write_handler, output_repr);
+
+  if (
+    !yaml_stream_start_event_initialize(&event, YAML_UTF8_ENCODING) ||
+    !yaml_emitter_emit(&emitter, &event)) {goto error;}
+  if (
+    !yaml_document_start_event_initialize(&event, NULL, NULL, NULL, 1) ||
+    !yaml_emitter_emit(&emitter, &event)) {goto error;}
+
+  if (!emit_type_description(&emitter, type_description)) {goto error;}
+
+  if (
+    !yaml_document_end_event_initialize(&event, 1) ||
+    !yaml_emitter_emit(&emitter, &event)) {goto error;}
+  if (
+    !yaml_stream_end_event_initialize(&event) ||
+    !yaml_emitter_emit(&emitter, &event)) {goto error;}
+
+  yaml_emitter_delete(&emitter);
+  // Remove null-terminator character which should not be part of hashed content
+  output_repr->buffer_length--;
```

Comment:
I'm not sure that I agree with doing this here.  For one thing, this likely messes up the `rcutils_char_array` internal bookkeeping.  For another, this is probably surprising to external users of this API (i.e. not `rcl_calculate_type_version_hash`).

My suggestion here is to leave the `rcutils_char_array` as-is in this function, and deal with removing the `\0` in `rcl_calculate_type_version_hash`.

Reply:
I will note that it does not mess up the internal bookkeeping - there are two values, `buffer_capacity`, and `buffer_length`. Length marks how long the content currently is and can be manipulated arbitrarily, so long as it is not made longer than the capacity. The capacity on the other hand, needs to be carefully modified via APIs. There are no such APIs for modification of the length.

However, it's fine with me to shift the responsibility - that's how I had it in my first draft. The end user just has to know to use either `strlen(msg_repr.buffer`), or `msg_repr.buffer_length - 1` for the size of data to hash. I will note it in the docstring